services/worker-manager/src/providers/aws.js

```diff
@@ -0,0 +1,168 @@
+const {Provider} = require('./provider');
+const aws = require('aws-sdk');
+const _ = require('lodash');
+
+class AwsProvider extends Provider {
+  constructor({WorkerPool, instancePermissions, apiVersion, region}) {
+    super({WorkerPool});
+    this.configSchema = 'config-aws';
+    this.instancePermissions = instancePermissions;
+    this.apiVersion = apiVersion;
+    this.region = region;
+
+    aws.config.update({region});
+    this.iam = new aws.IAM({apiVersion});
+  }
+
+  /*
+   This method is used to setup permissions for the EC2 instances
+   */
+  async setup() {
+
+    // POLICY
+    const ec2policyName = 'TaskclusterWorkerAWSAccess';
+    const ec2PolicyArn = `arn:aws:iam::aws:policy/${ec2policyName}`;
+    const ec2policy = {
+      "Statement": [{
+        "Version": "2012-10-17",
+        "Effect": "Allow",
+        "Action": this.instancePermissions.actions,
+        "Resource": this.instancePermissions.resource,
+      }],
+    };
+
+    const freshPolicyMetadata = await readModifySet({
+      read: async () => await this.iam.getPolicy({PolicyArn: ec2PolicyArn}),
+      compare: async policyMetadata => {
+        const existingPolicy = await this.iam.getPolicyVersion({
+          PolicyArn: ec2PolicyArn,
+          VersionId: policyMetadata.DefaultVersionId,
+        });
+        const existingPolicyDocument = JSON.parse(existingPolicy.Document);
+
+        return _.isEqual(existingPolicyDocument.Statement, ec2policy.Statement);
+      },
+      modify: async policyMetadata => {
+        // This won't let us roll back to a different version, but the code is simpler
+        // possibly a temporary way of modifying a policy
+        // a way that would allow to roll back:
+        // listPolicyVersions({PolicyArn}) --> sort them by creation date --> delete the oldest --> createPolicyVersion
+        // there can only be 5 versions of a policy.
+        await this.iam.deletePolicyVersion({
```

Comment:
Another option we can take here is to say that this is not modifiable. If that is easier and makes this make more sense, we can take that route.

Reply:
Would that option disallow users add or remove instance permissions? (i.e. those would be set once and for all?)

Reply:
(Another possible thing to do here is to delete a policy entirely, but in order to do that, it must be detached from everywhere.....)

Reply:
If testing shows that this approach works reliably, let's keep it!  If not, I think not allowing modification is a good option too.  Currently, most of our instances don't have a role at all.  The exception is OCC-controlled Windows instances, and [that is changing](https://bugzilla.mozilla.org/show_bug.cgi?id=1562686).